More cli sanity verifications

fabianofranz opened the pull request:
Adds some more `kubectl` command sanity checks to improve consistency and avoid the need of code reviews for some of our CLI style and standards.

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access) 
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`. 
-->
```release-note
NONE
```
@kubernetes/sig-cli-pr-reviews

fabianofranz committed:
More cli sanity verifications

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/45308)
<!-- Reviewable:end -->

adohe commented on pkg/kubectl/cmd/util/sanity/cmd_sanity.go:
```diff
@@ -92,3 +108,45 @@ func CheckExamples(cmd *cobra.Command) []error {
+		name := flag.Name
+
+		regex := `^[a-z]+[a-z\-]*$`
+		matches, err := regexp.MatchString(regex, name)
```
what about compiling the regex only once? not compile this every flag.

adohe commented:
Overall this looks good to me. @fabianofranz is there any doc about what cli conventions check does? I cannot find one.

adohe commented:
@eparis ptal.

fabianofranz commented:
@adohe thanks, fixed the regex compilation. No docs about this specific `verify` script, I'll add a paragraph to [kubectl-conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/kubectl-conventions.md) in a separate PR.

fabianofranz commented:
/assign @dchen1107

alexandercampbell commented:
Review status: 0 of 4 files reviewed at latest revision, 4 unresolved discussions.

---

*[cmd/clicheck/check_cli_conventions.go, line 40 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45308#-KkO-5diUCoKDNxAMBqF:-KkO-5djMrmI1w_rg7oH:b-h4juwx) ([raw file](https://github.com/kubernetes/kubernetes/blob/791a43a2e601e6066042b131fbc3021f812b4c21/cmd/clicheck/check_cli_conventions.go#L40)):*
> ```Go
> 	for _, err := range errors {
> 		errorCount++
> 		fmt.Fprintf(os.Stderr, "     %d. %s\n", errorCount, err)
> ```

Would it make more sense to use `%5d` or similar here?

---

*[pkg/kubectl/cmd/util/sanity/cmd_sanity.go, line 45 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45308#-KkO-bLhrVZGyrjcvtPr:-KkO-bLhrVZGyrjcvtPs:b-lkrktt) ([raw file](https://github.com/kubernetes/kubernetes/blob/791a43a2e601e6066042b131fbc3021f812b4c21/pkg/kubectl/cmd/util/sanity/cmd_sanity.go#L45)):*
> ```Go
> 
> func RunGlobalChecks(globalChecks []GlobalCheck) []error {
> 	fmt.Fprintf(os.Stdout, "---+ RUNNING GLOBAL CHECKS\n")
> ```

Can just be `fmt.Print` unless (or is `fmt.Print` being redirected?)

---

*[pkg/kubectl/cmd/util/sanity/cmd_sanity.go, line 145 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45308#-KkO0D2zQP1Gu2MInl7Y:-KkO0D2zQP1Gu2MInl7Z:bkkbul2) ([raw file](https://github.com/kubernetes/kubernetes/blob/791a43a2e601e6066042b131fbc3021f812b4c21/pkg/kubectl/cmd/util/sanity/cmd_sanity.go#L145)):*
> ```Go
> 	fmt.Fprintf(os.Stdout, "   ↳ checking flags from global vars\n")
> 	errors := []error{}
> 	if pflag.CommandLine.HasFlags() {
> ```

Isn't this conditional redundant? `VisitAll` should be a no-op for zero flags.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45308)*
<!-- Sent from Reviewable.io -->

fabianofranz commented:
Review status: 0 of 4 files reviewed at latest revision, 4 unresolved discussions.

---

*[cmd/clicheck/check_cli_conventions.go, line 40 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45308#-KkO-5diUCoKDNxAMBqF:-KkRs9TwEbh37W27X5A1:btls8e6) ([raw file](https://github.com/kubernetes/kubernetes/blob/791a43a2e601e6066042b131fbc3021f812b4c21/cmd/clicheck/check_cli_conventions.go#L40)):*
<details><summary><i>Previously, alexandercampbell (Alexander Campbell) wrote…</i></summary><blockquote>

Would it make more sense to use `%5d` or similar here?
</blockquote></details>

In fact no, since I need fixed spaces width on the left.

---

*[pkg/kubectl/cmd/util/sanity/cmd_sanity.go, line 131 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45308#-KjFC0ZNL9BkmulgXxjG-r1-131:-KkRsI_klx2Ck3pmfFHk:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/b239b4404e4f7cb1653165feb16b09c675710a6a/pkg/kubectl/cmd/util/sanity/cmd_sanity.go#L131)):*
<details><summary><i>Previously, adohe (TonyAdo) wrote…</i></summary><blockquote>

what about compiling the regex only once? not compile this every flag.
</blockquote></details>

Done.

---

*[pkg/kubectl/cmd/util/sanity/cmd_sanity.go, line 45 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45308#-KkO-bLhrVZGyrjcvtPr:-KkRsEurCH5MG5ihuyI2:be01dwz) ([raw file](https://github.com/kubernetes/kubernetes/blob/791a43a2e601e6066042b131fbc3021f812b4c21/pkg/kubectl/cmd/util/sanity/cmd_sanity.go#L45)):*
<details><summary><i>Previously, alexandercampbell (Alexander Campbell) wrote…</i></summary><blockquote>

Can just be `fmt.Print` unless (or is `fmt.Print` being redirected?)
</blockquote></details>

It can be `fmt.Fprintf`, fixed.

---

*[pkg/kubectl/cmd/util/sanity/cmd_sanity.go, line 145 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45308#-KkO0D2zQP1Gu2MInl7Y:-KkRsJSjfKjX5rqGAM7R:b8rko35) ([raw file](https://github.com/kubernetes/kubernetes/blob/791a43a2e601e6066042b131fbc3021f812b4c21/pkg/kubectl/cmd/util/sanity/cmd_sanity.go#L145)):*
<details><summary><i>Previously, alexandercampbell (Alexander Campbell) wrote…</i></summary><blockquote>

Isn't this conditional redundant? `VisitAll` should be a no-op for zero flags.
</blockquote></details>

Good catch , fixed.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45308)*
<!-- Sent from Reviewable.io -->

fabianofranz commented:
Comments addressed, @alexandercampbell thanks for the review! PTAL

fabianofranz commented:
Review status: 0 of 4 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/kubectl/cmd/util/sanity/cmd_sanity.go, line 45 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45308#-KkO-bLhrVZGyrjcvtPr:-KkRwI_gVX7G5vUt-k6p:b-6w3vsk) ([raw file](https://github.com/kubernetes/kubernetes/blob/791a43a2e601e6066042b131fbc3021f812b4c21/pkg/kubectl/cmd/util/sanity/cmd_sanity.go#L45)):*
> fmt.Fprintf

`fmt.Fprint` actually

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45308)*
<!-- Sent from Reviewable.io -->

alexandercampbell commented:
/lgtm

---







*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45308#-:-KkSM1uhuRD5mbKZbc9u:b-8ks5qa)*
<!-- Sent from Reviewable.io -->

fabianofranz commented:
@adohe mind lgtm'ing?
